File: Esercitazioni/Esercitazione0/svolta/produttore.c
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#define MAX_STRING_LENGTH 256

// produttore.c NON e' un filtro
int main(int argc, char *argv[]) {
    int   fd, readValues, bytes_to_write, written, righe, i;
    char *file_out;
    char  riga[MAX_STRING_LENGTH], buf[MAX_STRING_LENGTH];

    // controllo numero argomenti
    if (argc != 2) {
        perror(" numero di argomenti sbagliato");
        exit(1);
    }

    file_out = argv[1];
    printf("Quante righe vuoi inserire?\n");
    readValues = scanf("%d", &righe);
    if (readValues != 1) {
        printf("Errore: immettere un intero!");
        exit(1);
    }
    gets(buf); // consumare il fine linea

    fd = open(file_out, O_WRONLY | O_CREAT | O_TRUNC, 00640);
    if (fd < 0) {
        perror("P0: Impossibile creare/aprire il file");
        exit(2);
    }

    for (i = 0; i < righe; ++i) {
        printf("Inserisci la nuova riga\n");
        gets(riga);
        /* la gets legge tutta la riga, separatori inclusi, e trasforma il fine
           linea in fine stringa */
        riga[strlen(riga)]     = '\n'; // aggiungo il fine linea
        riga[strlen(riga) + 1] = '\0'; // aggiungo 0 binario per permettere una corretta strlen()
        written = write(fd, riga, strlen(riga)); // uso della primitiva
        if (written < 0) {
            perror("P0: errore nella scrittura sul file");
            exit(3);
        }
    }
    close(fd);
}

File: Esercitazioni/Esercitazione0/svolta/Produttore.java
import java.io.BufferedReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.UnsupportedEncodingException;

// Produttore NON e' un filtro
public class Produttore {
	public static void main(String[] args) {		
		BufferedReader in = null;
		int res = 0;
		
		if (args.length != 1){
			System.out.println("Utilizzo: produttore <inputFilename>");
			System.exit(0);
		}
		
		System.out.println("Quante righe vuoi inserire?");
		in = new BufferedReader(new InputStreamReader(System.in));
			
		FileWriter fout;
		String inputl = null;
		try {
			fout = new FileWriter(args[0]);
			res = Integer.parseInt(in.readLine());
			for (int i =0; i<res; i++){
				System.out.println("Inserisci la nuova riga");
				inputl = in.readLine()+"\n";
				fout.write(inputl, 0, inputl.length());
			}		
			fout.close();
		} 
		catch (NumberFormatException nfe) { 
			nfe.printStackTrace(); 
			System.exit(1); // uscita con errore, intero positivo a livello di sistema Unix
		}
	    catch (IOException e) { 
			e.printStackTrace();
			System.exit(2); // uscita con errore, intero positivo a livello di sistema Unix
		}
	}
}

